Add support for extending configurations

rexxars opened the pull request:
This PR adds support for the `extends` configuration option, which allows users to extend other configurations.

The primary use case I want to address is where you have a coding standard (for instance as an npm package) and want to make small modifications to it, for instance adding plugins or similar. In this case, cascading configurations doesn't help.

The PR addresses #1637, where @nzakas said:
> If you want to implement extends, I'd say go for it.

This is my first contribution, feedback is appreciated :-)

ilyavolodin commented on lib/config.js:
```diff
@@ -61,6 +62,16 @@ function loadConfig(filePath) {
+            config.extends
+        );
+
+        config = util.mergeConfigs(loadConfig(parentPath), config);
```
I wonder if there should be a try-catch block around this? It would be nice to have an error message specify which config file cause this problem (or rather which one was the root of the problem).

rexxars replied:
I suppose you mean in case the `extends` file specified could not be found/parsed?
Could do that, but since it's recursive it's a bit tricky. What do we want the output to be?

We could, for instance, add the source configs to the message, so the output would be something like:

```
Cannot read config file: /home/espenh/webdev/eslint/tests/fixtures/config-extends/subdir/subsubdir/subsubsubdir/deepesta.json
Error: ENOENT: no such file or directory, open '/home/espenh/webdev/eslint/tests/fixtures/config-extends/subdir/subsubdir/subsubsubdir/non-existing-file.json'
Referenced from: /home/espenh/webdev/eslint/tests/fixtures/config-extends/subdir/subsubdir/deeper.json
Referenced from: /home/espenh/webdev/eslint/tests/fixtures/config-extends/deep.json
```
Or possibly something a little prettier/better. Suggestions?

ilyavolodin replied:
Yes, adding file path to the error and then combining all of them into something meaningful would probably be a good idea. I think the message above is fine. It's descriptive enough so that you could fine and fix an issue.

nzakas replied:
Won't this already throw an error if the referenced file isn't found?

rexxars replied:
Yup, it does. But there's no real clue telling where that file was referenced from. The above example shows the "entry config" and the whole chain of extends.

nzakas commented on lib/config.js:
```diff
@@ -20,7 +20,8 @@ var fs = require("fs"),
     yaml = require("js-yaml"),
-    userHome = require("user-home");
+    userHome = require("user-home"),
+    isAbsolutePath = require("path-is-absolute");
```
Do we need a new module for this? Isn't this just checking to see if the first character is `/`?

rexxars replied:
I hated adding this, but if we want to support windows, we unfortunately can't just check for /. Node 0.12 has a native `path.isAbsolute()`, but contribution guidelines says to still support 0.10, so...

sindresorhus replied:
https://github.com/sindresorhus/path-is-absolute

rexxars replied:
... that is what we're using :p

nzakas commented on lib/config.js:
```diff
@@ -61,6 +62,16 @@ function loadConfig(filePath) {
 
+    // Recursively merge parent configurations
+    if (config.extends) {
+        var parentPath = (!isAbsolutePath(config.extends) ?
```
Can you add a comment explaining what this is doing?

rexxars commented:
- Added more meaningful stack trace when referenced configuration could not be loaded/parsed (suggested by @ilyavolodin)
- Added more comments to explain each step of the process
- Squashed and pushed

Please let me know if this needs more work.

nzakas commented on docs/user-guide/configuring.md:
```diff
@@ -391,6 +391,21 @@ The complete configuration hierarchy, from highest precedence to lowest preceden
 
+## Extending Configuration Files
+
+If you want to extend an existing configuration file which is not located in a parent directory, you can use the `extends` field to specify the location. For example:
```
This is really any file, including one in a parent directory.

We probably also need to mention that the extended config provides base rules that are overridden by this config.

nzakas commented:
This looks really good, I think we just need a bit more documentation.

rexxars commented:
Made the documentation a bit clearer. If there are any specific things you want improved in the documentation, I'd love some help on it. Though the theory on how this feature works is fairly simple, English is not my primary language and I find explaining this stuff a bit challenging :p

nzakas commented:
The docs look good to me. Thanks!
